graders/python/python_autograder/code_feedback.py

```diff
@@ -117,6 +127,24 @@ def finish(cls, fb_text: str) -> NoReturn:
         cls.add_feedback(fb_text)
         raise GradingComplete("Your answer is correct.")
 
+    @classmethod
+    def finish_test(cls, fb_text: str) -> NoReturn:
```

Comment:
This function name isn't great. It could be `abort_test` or `exit_test`. `finish_test` implies other meanings I feel.

Reply:
I'm not too worried about naming since we're going to rework the autograder at some point anyway.

Reply:
There isn't actually anything inherently "unhappy" about calling `finish_test`, just like for `finish`. It just says "halt execution". Whether that's positive or negative is determined by if the test set points or provided error feedback or whatever.

That said, that's not clear from the name of the exception (`GradingTestFailedError`). I do wonder if we're making `finish_test` a variant of `finish`, maybe we should apply the same pattern to `GradingComplete` and name it `GradingTestComplete` (or `GradingTestCompleteError` if we want to make Ruff happy)?

I acknowledge that we'll rewrite the Python grader at some point, but we've also been saying that for many years, so I'd rather we continue to treat the current Python grader with the same standards we use for all our other code.

Reply:
I think `GradingTestComplete` is a reasonable name. I'll add an ignore comment since this is an example where the `Error` suffix isn't very descriptive (and generally a symptom of the fact that using exceptions for regular control flow isn't great).

Reply:
Agreed that using exceptions for control flow is generally messy, though it's definitely powerful when used judiciously (we use it in some parts of PL, it lets you build some pretty nice APIs).

Reply:
The main argument I have seen against Exceptions is that it you can't easily look at a function and see what Exceptions it can throw, and the control flow can be less intuitive. I do think it can help avoid other patterns, like the Go pattern `val, ok :=` for error propagation creating a lot of boilerplate code. Rust has the nice [`do_thing()?` macro](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator), but python has no such convention or pattern for this.